experimental/internal/taxa/taxa.go

```diff
@@ -0,0 +1,217 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// package taxa (plural of taxon, an element of a taxonomy) provides support for
+// classifying Protobuf syntax productions for use in the parser and in
+// diagnostics.
+//
+// The Subject enum is also used in the parser stack as a simple way to inform
+// recursive descent calls of what their caller is, since the What enum
+// represents "everything" the parser stack pushes around.
+package taxa
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/bufbuild/protocompile/internal/iter"
+)
+
+// Subject is a syntactic or semantic element within the grammar that can be
+// referred to within a diagnostic.
+type Subject int
+
+const (
+	Unknown Subject = iota
+	Unrecognized
+	TopLevel
+	EOF
+
+	Decl
+	Empty
+	Syntax
+	Edition
+	Package
+	Import
+	WeakImport
+	PublicImport
+	Extensions
+	Reserved
+	Body
+
+	Def
+	Message
+	Enum
+	Service
+	Extend
+	Oneof
+
+	Option
+	CustomOption
+
+	Field
+	EnumValue
+	Method
+
+	FieldTag
+	OptionValue
+
+	CompactOptions
+	MethodIns
+	MethodOuts
+
+	Path
+	ExtensionInPath
+
+	Expr
+	Range
+	Array
+	Dict
+	DictField
+
+	Type
+	TypeParams
+
+	Whitespace
+	Comment
+	Ident
+	String
+	Float
+	Int
+
+	Semicolon
+	Comma
+	Slash
+	Colon
+	Equals
+	Minus
+	Period
+
+	LParen
+	LBracket
+	LBrace
+	LAngle
+
+	RParen
+	RBracket
+	RBrace
+	RAngle
+
+	Parens
+	Brackets
+	Braces
+	Angles
+
+	KeywordSyntax
+	KeywordEdition
+	KeywordImport
+	KeywordWeak
+	KeywordPublic
+	KeywordPackage
+
+	KeywordOption
+	KeywordMessage
+	KeywordEnum
+	KeywordService
+	KeywordExtend
+	KeywordOneof
+
+	KeywordExtensions
+	KeywordReserved
+	KeywordTo
+	KeywordRPC
+	KeywordReturns
+
+	KeywordOptional
+	KeywordRepeated
+	KeywordRequired
+	KeywordGroup
+	KeywordStream
+
+	// Total is the total number of known [What] values.
+	Total int = iota
```

Comment:
This seems like it should _not_ be exported.
```suggestion
	// total is the total number of known [Subject] values.
	Total int = iota
```